### ISSUE TITLE
TensorRT Engine on CUDA:1 Fails with invalid resource handle During Inference

### ISSUE THREAD

**Issue by CongSuxu** [ultralytics/ultralytics#21300]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I exported two TensorRT engines from the same model, targeting different GPUs:

```python
# Export to CUDA:0
model = YOLO('yolo11s.pt').to('cuda:0')
device = torch.device('cuda:0')
model.export(format='engine', batch=64, device=device)

# Export to CUDA:1
model = YOLO('yolo11s.pt').to('cuda:1')
device = torch.device('cuda:1')
model.export(format='engine', batch=64, device=device)

# Rename outputs:
# - yolo11s_0.engine  -> for CUDA:0
# - yolo11s_1.engine  -> for CUDA:1
````
Then I tested inference:
```python
model = YOLO('yolo11s_1.engine')
img = torch.zeros(64, 3, 640, 640, dtype=torch.float32).to('cuda:1')
result = model.predict(img)
```
When using yolo11s_0.engine on cuda:0, there's no error. When using yolo11s_1.engine on cuda:1, It fails with the following error:
```
[07/07/2025-20:34:31] [TRT] [E] IExecutionContext::executeV2: Error Code 1: CuTensor (Internal cuTensor permutate execute failed)
[07/07/2025-20:34:31] [TRT] [E] [checkMacros.cpp::nvinfer1::catchCudaError::212] Error Code 1: Cuda Runtime (invalid resource handle)
```

### Environment

Ultralytics 8.3.162  Python-3.9.21 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3090, 24576MiB)
Setup complete  (20 CPUs, 31.8 GB RAM, 146.0/465.1 GB disk)

OS                  Windows-10-10.0.19045-SP0
Environment         Windows
Python              3.9.21
Install             git
Path                C:\Users\admin\Desktop\torchServe\system\python\Lib\site-packages\ultralytics
RAM                 31.84 GB
Disk                146.0/465.1 GB
CPU                 Intel Core(TM) i5-14600KF
CPU count           20
GPU                 NVIDIA GeForce RTX 3090, 24576MiB
GPU count           2
CUDA                12.1

numpy                1.26.3>=1.23.0
matplotlib           3.9.4>=3.3.0
opencv-python        4.6.0.66>=4.6.0
pillow               11.0.0>=7.1.2
pyyaml               6.0.2>=5.3.1
requests             2.32.3>=2.23.0
scipy                1.13.1>=1.4.1
torch                2.4.1+cu121>=1.8.0
torch                2.4.1+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision          0.19.1+cu121>=0.9.0
tqdm                 4.67.1>=4.64.0
psutil               7.0.0
py-cpuinfo           9.0.0
pandas               2.2.3>=1.1.4
ultralytics-thop     2.0.14>=2.0.0

### Minimal Reproducible Example

on the  above.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @CongSuxu, thank you for your detailed report and for using Ultralytics 🚀! This is an automated response to help you get started. An Ultralytics engineer will review your issue and assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/), where you’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as answers to common questions.

Since this is a 🐛 Bug Report, please ensure you've provided a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/). It looks like you've included your code — thank you! If possible, please double-check that the example lets others reproduce the error from a clean environment.

For real-time support, join our [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussion, visit [Discourse](https://community.ultralytics.com/), or share insights on our [Subreddit](https://www.reddit.com/r/Ultralytics/).

## Upgrade

First, please upgrade to the latest `ultralytics` release and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). This ensures your issue hasn’t already been fixed:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your report! An Ultralytics engineer will be with you shortly to assist further.

**Comment by Y-T-G** [comment]
You can try passing the device in prediction.

```python
import torch

results = model.predict(..., device=torch.device(1))
```

**Comment by Y-T-G** [comment]
You can also check this on how to maintain two different CUDA contexts for two different GPUs in the same session


https://github.com/ultralytics/ultralytics/issues/17973#issuecomment-2516828801
